packages/project-access/src/project/ui5-config.ts

```diff
@@ -42,3 +47,98 @@ export async function readUi5Yaml(projectRoot: string, fileName: string, memFs?:
     }
     throw Error(`File '${fileName}' not found in project '${projectRoot}'`);
 }
+
+/**
+ * Scans the project directory for ui5 configuration yaml files.
+ *
+ * @param memFs - mem-fs editor instance
+ * @param projectRoot - path to project root, where ui5 configuration y*ml files are located
+ * @param validateSchema - (default: true) indicator to split list of file names into valid and invalid ones based on the ui5 configuration schema
+ * @returns list of valid and invalid UI5 configuration yaml file names
+ * @throws {Error} if an error occurs while reading files from projectRoot
+ */
+export async function getAllUi5YamlFileNames(
+    memFs: Editor,
+    projectRoot: string,
+    validateSchema = true
+): Promise<ValidatedUi5ConfigFileNames> {
+    return new Promise((resolve) => {
+        //use 'fs' here directly because we only create a list of file names without any i/o operations
+        readdir(projectRoot, async (error, files) => {
+            if (error) {
+                throw new Error(`Error reading files from directory '${projectRoot}': ${error}`);
+            }
+            const yamlFileNames = new Set<string>();
+            files
+                .filter((file) => file.endsWith('.yaml') || file.endsWith('.yml'))
+                .forEach((file) => {
+                    yamlFileNames.add(file);
+                });
+            //add not yet saved .y*ml files from mem-fs as well
+            const memYamlFiles = memFs.dump(
+                projectRoot,
+                (file) => file.basename.endsWith('.yaml') || file.basename.endsWith('.yml')
+            );
+            for (const file in memYamlFiles) {
+                if (memYamlFiles[file].state === 'deleted') {
+                    yamlFileNames.delete(file);
+                    continue;
+                }
+                yamlFileNames.add(file);
+            }
+            resolve(
+                validateSchema
+                    ? await validateUi5ConfigSchema(memFs, yamlFileNames, projectRoot)
+                    : ({
+                          valid: Array.from(yamlFileNames)
+                      } satisfies ValidatedUi5ConfigFileNames)
+            );
+        });
+    });
+}
+
+/**
+ * Validates the schema of the given yaml files.
+ * Multi-document sources are currently not supported and will be skipped.
+ *
+ * @param memFs - mem-fs editor instance
+ * @param yamlFileNames - list of yaml file names to be validated
+ * @param projectRoot - path to project root, where ui5 configuration y*ml files are located
+ * @returns list of valid and invalid UI5 configuration yaml file names
+ * @throws {Error} if the schema for the validation cannot be read
+ */
+export async function validateUi5ConfigSchema(
```

Comment:
Context: Currently the converter reads the yaml file names via project-access and get back a validated (also via project access) list of all  yaml files (valid, invalid, skipped).

Did I get you right that you suggest to still get the list of yaml files from project-access but move tha validation of the content to a later point in time, namely when the single config file is being read?

For the converter this would mean that it doesn't get a list of pre-validated files upfront and can exclude those files (currently done in `isUi5YamlToBeConverted` in ui5-yaml.ts) but needs to try to open each single file to trigger a validation (currently done in `processUi5YamlConfig` in ui5.yaml.ts).

So I guess for the converter it doesn't really matter but I'm not sure what this means performance wise. Can we share the schema validator instance across multiple ui5 yaml file config instances? We should not have one instance per file as the schema is anyway the same for all of them and the schema file is huge so I guess the validator instance is also quite heavy.

As far as I know ui5-config currently only comes with single file operations. We could as well add some multi-file operations (schema validating a list of config files would be the first one, not sure if there will ever be more). That would basically just be moving the schema validation function as is from project-access to ui5-config.

Reply:
How about enhancing class `UI5Config` with a private static member that is basically a singleton of the validate function. Instantiation is lazy (first validation request on instance level would be a bit slower because it also triggers the instantiation of the validate function). Validation can then be triggered if needed e.g. when creating a `newInstance` but all instances share the same validate function.

```ts
export class UI5Config {
    private document: YamlDocument;
    private static validate: ValidateFunction<SomeJSONSchema>; // <----- singleton validate function

    /**
     * Validates the schema of the given yaml document.
     *
     * @returns true if the document is valid, false otherwise
     */
    async validateSchema(): Promise<boolean> {
        if (!UI5Config.validate) { // <----- lazy loading
            const schema = JSON.parse(
                await readFile(join(__dirname, '..', '..', 'dist', 'schema', 'ui5.yaml.json'), 'utf8')
            ) as SomeJSONSchema | null;
            if (!schema) {
                throw Error('Schema not found. No validation possible.');
            }
            UI5Config.validate = new Ajv({ strict: false }).compile<SomeJSONSchema>(schema);
        }

        return UI5Config.validate(document); // <----- actual trigger of the validate function
    }

    /**
     * Returns a new instance of UI5Config.
     *
     * @static
     * @param {string} serializedYaml - the serialized yaml string
     * @param options - options
     * @param [options.validateSchema] - whether to validate the schema of the yaml file
     * @returns {UI5Config} the UI5Config instance
     * @throws {Error} if the schema validation fails
     * @memberof UI5Config
     */
    static async newInstance(
        serializedYaml: string,
        options?: {
            validateSchema?: boolean;
        }
    ): Promise<UI5Config> {
        const instance = new UI5Config();
        instance.document = await YamlDocument.newInstance(serializedYaml);
        if (options?.validateSchema) {
            const isValid = await instance.validateSchema();
            if (!isValid) {
                throw new Error('Invalid UI5 configuration file');
            }
        }
        return instance;
    }

    //...
```